chia/data_layer/data_store.py

```diff
@@ -58,26 +51,52 @@ async def create(cls, db_wrapper: DBWrapper) -> "DataStore":
 
         async with self.db_wrapper.locked_transaction():
             await self.db.execute(
-                """
+                f"""
                 CREATE TABLE IF NOT EXISTS node(
-                    hash TEXT PRIMARY KEY NOT NULL,
-                    node_type INTEGER NOT NULL,
-                    left TEXT REFERENCES node,
-                    right TEXT REFERENCES node,
-                    key TEXT,
-                    value TEXT
+                    hash BLOB PRIMARY KEY NOT NULL CHECK(length(hash) == 32),
+                    node_type INTEGER NOT NULL CHECK(
+                        (
+                            node_type == {int(NodeType.INTERNAL)}
+                            AND left IS NOT NULL
+                            AND right IS NOT NULL
+                            AND key IS NULL
+                            AND value IS NULL
+                        )
+                        OR
+                        (
+                            node_type == {int(NodeType.TERMINAL)}
+                            AND left IS NULL
+                            AND right IS NULL
+                            AND key IS NOT NULL
+                            AND value IS NOT NULL
+                        )
+                    ),
+                    left BLOB REFERENCES node,
+                    right BLOB REFERENCES node,
+                    key BLOB,
+                    value BLOB
                 )
                 """
             )
             await self.db.execute(
                 """
+                CREATE TRIGGER IF NOT EXISTS no_node_updates
+                BEFORE UPDATE ON node
```

Comment:
why do we need to prevent updates to the node table?

Reply:
When would you want to update a row?  I think it is an inappropriate activity for this store.  A node with different children is a different node and modifying an existing row accordingly would break references.  Or maybe that's your point?  Other constraints already protect against this?

Reply:
im confused as to what happens in "our" data_store when a row in the app is update with new values

Reply:
i was under the impression that a node represents a row so what happens when the user updates a row ? doesn't that change the node ?

Reply:
There's no prescribed mapping from their rows to our nodes.  But, we presently do not delete or modify nodes ever.  If you had a terminal/kv-pair node for `b"a key"`: `b"some value"` then it must be kept around forever so that it is there and accurate when older versions of the tree are requested.  If the user application deletes it and then creates a new terminal node such as `b"a key"`: `b"totally different"` then we create a new node with the same key, a different value, and a different hash.  New internal nodes are created all the way up to the root so that the new root hash has a path of descendants to the new terminal node with the new value instead of the old one.  The old root still retains all its descendants including the old terminal node with the old value.